packages/protocol/sybilprotection/activitytracker/activitytracker.go

```diff
@@ -0,0 +1,87 @@
+package activitytracker
+
+import (
+	"sync"
+	"time"
+
+	"github.com/iotaledger/hive.go/core/generics/options"
+	"github.com/iotaledger/hive.go/core/generics/shrinkingmap"
+	"github.com/iotaledger/hive.go/core/identity"
+	"github.com/iotaledger/hive.go/core/timed"
+
+	"github.com/iotaledger/goshimmer/packages/core/validator"
+)
+
+// TimeRetrieverFunc is a function type to retrieve the time.
+type TimeRetrieverFunc func() time.Time
+
+// region ActivityTracker //////////////////////////////////////////////////////////////////////////////////////////
+
+// ActivityTracker is a component that keeps track of active nodes based on their time-based activity in relation to activeTimeThreshold.
+type ActivityTracker struct {
+	timedExecutor     *timed.TaskExecutor[identity.ID]
+	lastActiveMap     *shrinkingmap.ShrinkingMap[identity.ID, time.Time]
+	timeRetrieverFunc TimeRetrieverFunc
+	validatorSet      *validator.Set
+
+	optsWorkersCount   uint
+	optsActivityWindow time.Duration
+
+	mutex sync.RWMutex
+}
+
+// New creates and returns a new instance of an ActivityTracker.
+func New(validatorSet *validator.Set, timeRetrieverFunc TimeRetrieverFunc, opts ...options.Option[ActivityTracker]) (activityTracker *ActivityTracker) {
+	return options.Apply(&ActivityTracker{
+		timeRetrieverFunc: timeRetrieverFunc,
+		validatorSet:      validatorSet,
+		lastActiveMap:     shrinkingmap.New[identity.ID, time.Time](),
+
+		optsWorkersCount:   1,
+		optsActivityWindow: time.Second * 30,
+	}, opts, func(a *ActivityTracker) {
+		a.timedExecutor = timed.NewTaskExecutor[identity.ID](int(a.optsWorkersCount))
+	})
+}
+
+// Update updates the underlying data structure and keeps track of active nodes.
+func (a *ActivityTracker) Update(activeValidator *validator.Validator, activityTime time.Time) {
+	a.mutex.RLock()
+	defer a.mutex.RUnlock()
```

Comment:
Why is this only a read lock? You are potentially modifying `lastActiveMap` and `validatorSet` in the same function call/context